### PR TITLE
Add a module "ecdf" (Empirical Cumulative Distribution Function) to "visualize" sub-package

### PR DESCRIPTION
**Pull request by DannieSheng**
**Describe your changes**
Add a new way to visualize: empirical cumulative distribution function (eCDF)
- A new function called "obj_size_ecdf" is added to the "visualize" sub-package

**Type of update**
Is this a:
* New feature
* Update to documentation
* Work in progress

Associated issues
#778

### CODE REVIEW

File: plantcv/plantcv/visualize/__init__.py [hunk 1]
```diff
@@ -6,6 +6,8 @@
 from plantcv.plantcv.visualize.auto_threshold_methods import auto_threshold_methods
 from plantcv.plantcv.visualize.overlay_two_imgs import overlay_two_imgs
 from plantcv.plantcv.visualize.colorize_label_img import colorize_label_img
+from . import ecdf
```
Comment (nfahlgren) [inline plantcv/plantcv/visualize/__init__.py:9]:
```suggestion
from plantcv.plantcv.visualize import ecdf
```

I think this is the import we want

File: plantcv/plantcv/visualize/ecdf/__init__.py [hunk 2]
```diff
@@ -0,0 +1,3 @@
+from .ecdf import obj_size, pix_intensity
```
Comment (nfahlgren) [inline plantcv/plantcv/visualize/ecdf/__init__.py:1]:
```suggestion
from plantcv.plantcv.visualize.ecdf.ecdf import obj_size, pix_intensity
```

File: plantcv/plantcv/__init__.py [hunk 3]
```diff
@@ -86,6 +86,7 @@
 from plantcv.plantcv.stdev_filter import stdev_filter
 from plantcv.plantcv.spatial_clustering import spatial_clustering
 from plantcv.plantcv import photosynthesis
+# from plantcv.plantcv.visualize import ecdf
```
Comment (nfahlgren) [inline plantcv/plantcv/__init__.py:89]:
```suggestion
```

File: plantcv/plantcv/visualize/__init__.py [hunk 4]
```diff
@@ -6,6 +6,8 @@
 from plantcv.plantcv.visualize.auto_threshold_methods import auto_threshold_methods
 from plantcv.plantcv.visualize.overlay_two_imgs import overlay_two_imgs
 from plantcv.plantcv.visualize.colorize_label_img import colorize_label_img
+from plantcv.plantcv.visualize.obj_size_ecdf import obj_size_ecdf
+
 
 __all__ = ["pseudocolor", "colorize_masks", "histogram", "clustered_contours", "colorspaces", "auto_threshold_methods",
-           "overlay_two_imgs", "colorize_label_img"]
+           "overlay_two_imgs", "colorize_label_img","pix_intensity","obj_size", "obj_size_ecdf"]
```
Comment (nfahlgren) [inline plantcv/plantcv/visualize/__init__.py:13]:
```suggestion
           "overlay_two_imgs", "colorize_label_img", "obj_size_ecdf"]
```

File: plantcv/plantcv/visualize/obj_size_ecdf.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
+# Plot Empirical Cumulative Distribution Function for Object Size
+
+import os
+import cv2
+import pandas as pd
+import numpy as np
```
Comment (nfahlgren) [inline plantcv/plantcv/visualize/obj_size_ecdf.py:6]:
```suggestion
```

Remove unused import

File: plantcv/plantcv/visualize/obj_size_ecdf.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+from plantcv.plantcv import params, fatal_error
```
Comment (nfahlgren) [inline plantcv/plantcv/visualize/obj_size_ecdf.py:7]:
```suggestion
from plantcv.plantcv import params
```

Remove unused import

File: plantcv/plantcv/visualize/obj_size_ecdf.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+from plantcv.plantcv._debug import _debug
+from statsmodels.distributions.empirical_distribution import ECDF
+from plotnine import ggplot, aes, geom_point, labels, \
+    scale_color_manual, scale_x_log10
```
Comment (nfahlgren) [inline plantcv/plantcv/visualize/obj_size_ecdf.py:11]:
```suggestion
from plotnine import ggplot, aes, geom_point, labels, scale_x_log10
```

Remove unused import

File: plantcv/plantcv/visualize/obj_size_ecdf.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    fig_ecdf = empirical cumulative distribution function plot
+
+    :param mask: numpy.ndarray
+    :param title: str
+    :return fig_ecdf: plotnine.ggplot.ggplot
+    """
+
+    objects, hierarchy = cv2.findContours(mask, cv2.RETR_TREE, cv2.CHAIN_APPROX_NONE)[-2:]
+    areas = [cv2.contourArea(cnt) for cnt in objects]
+
```
Comment (nfahlgren) [inline plantcv/plantcv/visualize/obj_size_ecdf.py:32]:
```suggestion
# Remove objects with areas < 1px
areas = [i for i in areas if i >= 1.0]

```

If we remove objects with areas < 1 it prevents RuntimeWarnings for log10

File: plantcv/plantcv/visualize/obj_size_ecdf.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    ecdf = ECDF(areas, side='right')
+
+    ecdf_df = pd.DataFrame({'object area': ecdf.x, 'cumulative probability': ecdf.y})
```
Comment (nfahlgren) [inline plantcv/plantcv/visualize/obj_size_ecdf.py:35]:
```suggestion
    ecdf_df = pd.DataFrame({'object area': ecdf.x[1:], 'cumulative probability': ecdf.y[1:]})
```

The eCDF starts with x=0 and y=-Inf, causing a warning in plotnine. If we exclude the first row then only valid values are plotted.

File: plantcv/plantcv/visualize/obj_size_ecdf.py [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+                + geom_point(size=.1)
+                + scale_x_log10())
+    if title is not None:
+        fig_ecdf = fig_ecdf + labels.ggtitle(title)
+
+    # Plot or print the ecdf
+    _debug(visual=fig_ecdf,
+           filename=os.path.join(params.debug_outdir, str(params.device) + '_area_ecdf.png'))
+    return fig_ecdf
+
```
Comment (nfahlgren) [inline plantcv/plantcv/visualize/obj_size_ecdf.py:47]:
```suggestion
```

Removes one extra line ending at end of file

File: tests/tests.py [hunk 11]
```diff
@@ -6384,9 +6384,19 @@ def test_plantcv_visualize_overlay_two_imgs_size_mismatch():
         _ = pcv.visualize.overlay_two_imgs(img1=img1, img2=img2)
 
 
+@pytest.mark.parametrize("title", ["Include Title", None])
+def test_plantcv_visualize_obj_size_ecdf(title, tmpdir):
```
Comment (nfahlgren) [inline tests/tests.py:6388]:
```suggestion
def test_plantcv_visualize_obj_size_ecdf(title):
```

tmpdir isn't needed as there's no file output by the test

File: tests/tests.py [hunk 12]
```diff
@@ -6384,9 +6384,19 @@ def test_plantcv_visualize_overlay_two_imgs_size_mismatch():
...
+    pcv.params.debug = None
+    cache_dir = tmpdir.mkdir("sub")
```
Comment (nfahlgren) [inline tests/tests.py:6390]:
```suggestion
```

File: docs/visualize_obj_size_ecdf.md [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+
+```python
+from plantcv import plantcv as pcv
+pcv.params.debug = "plot"
+fig_ecdf = pcv.visualize.ecdf.obj_size(mask=mask)
+```
+
+**Cumulative distribution of object sizes**
+
+![Screenshot](img/documentation_images/visualize_obj_size_ecdf/ecdf_obj_size.png)
```
Comment (nfahlgren) [inline docs/visualize_obj_size_ecdf.md:34]: If the areas < 1px are excluded the graph here will need to be regenerated.

**Review by nfahlgren — CHANGES_REQUESTED** [review]
Looks great and works great @DannieSheng, just a few suggestions below that would eliminate RuntimeWarnings and PlotnineWarnings

File: plantcv/plantcv/visualize/obj_size_ecdf.py (hunk 8 above)
Comment (DannieSheng) [reply plantcv/plantcv/visualize/obj_size_ecdf.py:30]: Good point!

**Review by nfahlgren — APPROVED** [review]

**Pull request merged**